From 59e2bf7250623fc17c9759e439d33299a3c1f923 Mon Sep 17 00:00:00 2001
From: Mian Yousaf Kaukab <yousaf.kaukab@suse.com>
Date: Thu, 11 May 2017 12:52:45 +0200
Subject: [PATCH] acpi: iort: thunderx2: enable smmu resource size errata for
 A1 silicon

Patch-mainline: Never, Workaround for early silicon
References: bsc#1035000

In next IORT spec release there will be a definition of a Cavium
specific model. Until then, enable the Cavium SMMU using cpu id
registers. Early silicon versions (A1) of Cavium's CN99xx SMMUv3
implementation must be enabled. For later silicon versions (B0) the
iort change will be in place.

Signed-off-by: Mian Yousaf Kaukab <yousaf.kaukab@suse.com>
---
 drivers/acpi/arm64/iort.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/drivers/acpi/arm64/iort.c b/drivers/acpi/arm64/iort.c
index 645073fb..a52f907b 100644
--- a/drivers/acpi/arm64/iort.c
+++ b/drivers/acpi/arm64/iort.c
@@ -766,6 +766,7 @@ static void __init arm_smmu_v3_init_resources(struct resource *res,
 	struct acpi_iort_smmu_v3 *smmu;
 	int num_res = 0;
 	unsigned long size = SZ_128K;
+	u32 cpu_model;
 
 	/* Retrieve SMMUv3 specific data */
 	smmu = (struct acpi_iort_smmu_v3 *)node->node_data;
@@ -773,8 +774,12 @@ static void __init arm_smmu_v3_init_resources(struct resource *res,
 	/*
 	 * Override the size, for Cavium ThunderX2 implementation
 	 * which doesn't support the page 1 SMMU register space.
+	 * For CN99xx A1 silicon override the size based on MIDR as
+	 * model is not correctly set in IORT
 	 */
-	if (smmu->model == ACPI_IORT_SMMU_CAVIUM_CN99XX)
+	cpu_model = read_cpuid_id() & MIDR_CPU_MODEL_MASK;
+	if (smmu->model == ACPI_IORT_SMMU_CAVIUM_CN99XX ||
+			cpu_model == 0x420f5160)
 		size = SZ_64K;
 
 	res[num_res].start = smmu->base_address;
-- 
2.11.0

